### PR TITLE
Kd_l231_cement_usa

### PR DESCRIPTION
**Pull request by kdorheim**
L231.cement USA translation complete, passes tests and checks. 

temp-inject stiching is required, having issues updating master to get outputs created by #789

### CODE REVIEW

File: R/zchunk_L2321.cement_USA.R [hunk 1]
```diff
@@ -68,114 +76,496 @@ module_gcam.usa_L2321.cement_USA_DISABLED <- function(command, ...) {
...
+      coefficient <- market.name <- grid_region <- stub.technology <- calibrated.value <-
+      tech.share.weight <- object <- NULL
+
+
     # ===================================================
-    # TRANSLATED PROCESSING CODE GOES HERE...
+    # Make the logit and input tables for with creating cement sectors in cement
+    # producing states.
+
+    # Not all states produce cement, save a vector of cement producing states bases on census data.
```
Comment (bpbond) [inline R/zchunk_L2321.cement_USA.R]: For clarity: "cement, so save"

File: R/zchunk_L2321.cement_USA.R [hunk 2]
```diff
@@ -68,114 +76,496 @@ module_gcam.usa_L2321.cement_USA_DISABLED <- function(command, ...) {
...
+      left_join_error_no_match(calibrated_techs_cement_sector_info, by = "sector") %>%
+      select(state, sector, calOutputValue, year, region, supplysector, subsector, technology) ->
+      L2321.StubTechProd_cement_USA
+
+    # Add share weight information to the state cement production data frame and format.
+    L2321.StubTechProd_cement_USA %>%
+      mutate(stub.technology = technology) %>%
+      mutate(share.weight.year = year) %>%
+      mutate(subs.share.weight = if_else(calOutputValue > 0, 1, 0)) %>%
+      mutate(tech.share.weight = subs.share.weight) %>%
```
Comment (bpbond) [inline R/zchunk_L2321.cement_USA.R]: Combine these four lines into a single `mutate` call

File: R/zchunk_L2321.cement_USA.R [hunk 3]
```diff
@@ -68,114 +76,496 @@ module_gcam.usa_L2321.cement_USA_DISABLED <- function(command, ...) {
...
+      select(region, supplysector, subsector, stub.technology, year, calOutputValue,
+             share.weight.year, subs.share.weight, tech.share.weight) ->
+      L2321.StubTechProd_cement_USA
+
+
+    # Create the coefficients of cement production technologies input table.
     #
+    # Start by creating a data frame of unique sector, fuel, supplysector, subsector,
+    # technology, and minicam.energy.input combinations. This data frame will be used to
+    # add sector information to input-ouput coefficients for state cement production.
```
Comment (bpbond) [inline R/zchunk_L2321.cement_USA.R]: Super commenting @kdorheim 👍

File: R/zchunk_L2321.cement_USA.R [hunk 4]
```diff
@@ -68,114 +76,496 @@ module_gcam.usa_L2321.cement_USA_DISABLED <- function(command, ...) {
...
-    # For more information, see https://github.com/JGCRI/gcamdata/wiki/Name-That-Function
-    # NOTE: This code uses vecpaste
-    # This function can be removed; see https://github.com/JGCRI/gcamdata/wiki/Name-That-Function
-    # NOTE: This code uses repeat_and_add_vector
-    # This function can be removed; see https://github.com/JGCRI/gcamdata/wiki/Name-That-Function
-    # ===================================================
+    # First change format of the the production default coefficients data frame from wide to long.
+    A321.globaltech_coef %>%
+      gather(year, value, matches(YEAR_PATTERN)) %>%
+      mutate(year = as.integer(year)) ->
```
Comment (bpbond) [inline R/zchunk_L2321.cement_USA.R]: Ideally we'll get #821 merged first and you can use `gather_years` here

File: R/zchunk_L2321.cement_USA.R [hunk 5]
```diff
@@ -68,114 +76,496 @@ module_gcam.usa_L2321.cement_USA_DISABLED <- function(command, ...) {
...
+      A321.globaltech_coef_long
+
+    # Then linearly interpolate the default coefficients for future years. In the next step these
+    # values will be added to the state input-output coefficients data frame.
+    A321.globaltech_coef_long %>%
+      complete(nesting(supplysector, subsector, minicam.energy.input, technology), year = c(year, FUTURE_YEARS)) %>%
+      arrange(supplysector, subsector, minicam.energy.input, technology, year) %>%
+      group_by(supplysector, subsector, minicam.energy.input, technology) %>%
+      mutate(value = approx_fun(year, value)) %>%
+      mutate(value = signif(value, energy.DIGITS_COEFFICIENT)) %>%
```
Comment (bpbond) [inline R/zchunk_L2321.cement_USA.R]: Again, single `mutate` call please

File: R/zchunk_L2321.cement_USA.R [hunk 6]
```diff
@@ -68,114 +76,496 @@ module_gcam.usa_L2321.cement_USA_DISABLED <- function(command, ...) {
...
+    # coefficients by supplysector / subsector / technology / minicam.energy.input combinations.
+    L2321.IO_GJkg_state_cement_F_Yh %>%
+      spread(year, value) %>%
+      left_join_error_no_match(L2321.globaltech_coef_future_years,
+                               by = c("supplysector", "subsector", "technology", "minicam.energy.input")) ->
+      IO_and_globaltech
+
+    # Format the the data frame and round the number of digits.
+    IO_and_globaltech %>%
+      gather(year, value, matches(YEAR_PATTERN)) %>%
```
Comment (bpbond) [inline R/zchunk_L2321.cement_USA.R]: `gather_years` hopefully; if not, combine `mutate`s

File: R/zchunk_L2321.cement_USA.R [hunk 7]
```diff
@@ -68,114 +76,496 @@ module_gcam.usa_L2321.cement_USA_DISABLED <- function(command, ...) {
...
+      filter(!supplysector %in% L2321.StubTechCoef_cement_USA$supplysector) ->
+      L2321.StubTechCalInput_cement_heat_USA_NOelectricity
+
+    # Add region and share weight information, format.
+    L2321.StubTechCalInput_cement_heat_USA_NOelectricity %>%
+      mutate(region = state) %>%
+      mutate(stub.technology = technology) %>%
+      mutate(share.weight.year = year) %>%
+      mutate(subs.share.weight = if_else(calibrated.value > 0, 1, 0)) %>%
+      mutate(tech.share.weight = subs.share.weight) %>%
```
Comment (bpbond) [inline R/zchunk_L2321.cement_USA.R]: (Here and throughout, combine mutates)

File: R/zchunk_L2321.cement_USA.R (hunk 4 above)
Comment (bpbond) [reply R/zchunk_L2321.cement_USA.R]: Replace these new lines with call to new `gather_year` function ( #821 ). This also applies to all `gather %>% mutate(year = as.integer(year))` sequences below.

**Comment by bpbond** [comment]
Waiting on @gokuliyer review.

File: R/zchunk_L2321.cement_USA.R [hunk 8]
```diff
@@ -68,114 +76,490 @@ module_gcam.usa_L2321.cement_USA_DISABLED <- function(command, ...) {
...
+    # input table.
+    calibrated_techs %>%
+      # We are only interested in the technology IDs where calibration = output.
+      filter(calibration == "output") %>%
+      select(sector, calibration, supplysector, subsector, technology) %>%
+      distinct ->
+      calibrated_techs_cement_sector_info
+
+    # Combine the cement sector information found above and the stub-technology calibrated
+    # cement production into a single data frame.
```
Comment (gokuliyer) [inline R/zchunk_L2321.cement_USA.R]: Do you mean tibble instead of data-frame?

File: R/zchunk_L2321.cement_USA.R (hunk 3 above)
Comment (gokuliyer) [reply R/zchunk_L2321.cement_USA.R]: Agreed.

**Pull request merged**